### DISCUSSION TITLE
Timezone within Neolink

### DISCUSSION THREAD

**Discussion by mirisbowring** [QuantumEntangledAndy/neolink#105] [Q&A]
Hi,

how can i set the correct timezone for my neolink installation?
When starting i get the following logs:
```
[2023-06-10T08:47:35Z INFO  neolink::rtsp::states::loggedin] K40: Camera time is already set: 2023-06-10 0:47:35.0 +08:00:00
[2023-06-10T08:47:35Z INFO  neolink::rtsp::states::loggedin] K40: Camera reports firmware version v2.0.0.354_19031100
[2023-06-10T08:47:35Z INFO  neolink::rtsp::states::streaming] K40: Starting video stream Main Stream (Clear)
[2023-06-10T08:47:35Z INFO  neolink::rtsp::states::streaming] K40: Starting video stream Sub Stream (Fluent)
```

But at the time, it is not 8:47 but 10:47 instead.

Thank you very much for this piece of Software!

**Comment by QuantumEntangledAndy** [comment]
Time zone is based on whatever your OS says it is.

So to answer this question I need to know what is your OS and if your using docker.

**Reply by mirisbowring** [reply]
Oh, the container is running on Unraid.
and `date` on the CLI prints the correct timestamp:
`Sat Jun 10 12:51:32 CEST 2023`

**Reply by QuantumEntangledAndy** [reply]
Set up the time zone on the container. there's an env variable that you can use. But I Can't remember what it's called atm. Best to just Google it. It's quite a common thing to do on a container and almost all my docker images have it set.

**Comment by nevilth** [comment]
Hi, I am in the process of setting up neolink on some Swann (reolink?) IPC_388 cameras and am having trouble with timezone also.

I am running docker container on debian 12 and am fairly new to it.

My docker-compose.yaml has the following:
neolink:
    container_name: neolink
    image: quantumentangledandy/neolink
    ports:
      - 8554:8554
    environment:
      - TZ=Australia/Melbourne
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /etc/timezone:/etc/timezone:ro
      - /opt/docker/neolink/config/neolink.toml:/etc/neolink.toml
    restart: unless-stopped

After starting I can view the logs by 

docker compose logs neolink

neolink  | Running: /bin/sh -c /usr/local/bin/neolink "${NEO_LINK_MODE}" --config /etc/neolink.toml
neolink  | [2024-08-15T07:21:47Z INFO  neolink] Neolink f15fdaefbcfe5cde3b0289c1e106cb85f6d3d39b release
neolink  | [2024-08-15T07:21:47Z WARN  neolink::rtsp] Without a server certificate, usernames and passwords will be exchanged in plaintext!
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Camera_4: Rtsp Starting
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Starting RTSP Server at 0.0.0.0:8554
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Camera_6: Rtsp Starting
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Camera_5: Rtsp Starting
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Camera_1: Rtsp Starting
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Camera_7: Rtsp Starting
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Camera_3: Rtsp Starting
neolink  | [2024-08-15T07:21:47Z INFO  neolink::rtsp] Camera_2: Rtsp Starting

System date:

nevil@THASRVR:/opt/docker$ date
Thu 15 Aug 2024 17:23:27 AEST

So the system date is set to AEST (Australia/Melbourne) time which is +10
The logs are being written with GMT.

Getting a bash shell in the container shows that the date time appears to be set correctly and the TZ variable is set.
 
nevil@THASRVR:/opt/docker$ docker compose run -it neolink bash
Running: bash

root@64b19fe89cff:/# date
Thu Aug 15 17:24:51 AEST 2024

root@64b19fe89cff:/# echo $TZ
Australia/Melbourne

Once the cameras come up, neolink sets the date /time as GMT.

neolink  | [2024-08-15T07:29:51Z WARN  neolink::common::camthread] Camera_3: Camera has no time set, Updating
neolink  | [2024-08-15T07:29:51Z INFO  neolink::common::camthread] Camera_3: Setting time to SystemTime { tv_sec: 1723706991, tv_nsec: 577389491 }
neolink  | [2024-08-15T07:29:51Z INFO  neolink::common::camthread] Camera_3: Camera time is now set: 2024-08-15 7:29:51.0 +00:00:00

Any ideas on what I am missing here?
And thank you for this software, its a life saver :-)

**Reply by QuantumEntangledAndy** [reply]
It's technically correct. The Z on the end of the means in standard time which is as you've noticed GMT.

It's not something I print myself I use (like most other programs) a library to handle the logging (specifically envlogger).

The envlogger library is what controls this and they have it set to use standard time unless you put in the effort to write a closure to override it. Which is not something I want since I'd prefer a consistent time zone in logs for debugging purposes.

**Reply by QuantumEntangledAndy** [reply]
Actually I've just noticed that you are not just talking about logs but the camera time. Hmmmm

**Reply by QuantumEntangledAndy** [reply]
Wait isn't the time still technically correct when set on the camera since that is at +0000 using gmt. The only issue is it's being printed to you in gmt, which should just be formatting again

**Reply by QuantumEntangledAndy** [reply]
Just spent the last 10 mins looking into this and since we use std::time to handle the time part of the code (which is not geared towards time display) it will always be in gmt.

If I wanted to fix this I'd probably just need to swap to the `time-rs` crate instead. Not sure it's something I'd want to do as it's just formatting, but maybe something to clean up if I'm ever editing that part.

If you want to PR it then the relevant part is in https://github.com/QuantumEntangledAndy/neolink/blob/f15fdaefbcfe5cde3b0289c1e106cb85f6d3d39b/src/common/camthread.rs#L197-L231

**Reply by nevilth** [reply]
Thanks for taking the time to look into it.
I take your point that it is just formatting and that the time is ultimately correct.
I will do a bit more investigation on my end and then maybe put a PR in.
